### ISSUE TITLE
Wasm, Contexts, and Entropy

### ISSUE THREAD

**Issue by JeremyRubin** [rust-bitcoin/rust-secp256k1#346]
> does signing context generate entropy on creation? testing now but this may be a big issue for using verification from a WASM module with no entropy because the get random number methods may panic.

For manually generated contexts, this happens only if you randomize them explicitly, see https://github.com/rust-bitcoin/rust-secp256k1/blob/48683d87c85ac28d5f2490c828979e7d8b9b874c/src/lib.rs#L354-L380. 

For the global context, it depends on the enabled feature: The `global-context` feature randomizes the context automatically and depends on `rand`, and  the `global-context-less-secure` feature gives you a context that is not randomized (and can't be randomized).

_Originally posted by @real-or-random in https://github.com/rust-bitcoin/rust-secp256k1/pull/342#issuecomment-991615743_

We should probably think through what it means to use any of these contexts in a WASM context! Especially in wasm32-unknown-unknown you have no syscalls or way of getting entropy, so this could be problematic.

Fully isolated WASM is really great to provide first-class support for for a myriad of reasons, but we'd need to think through carefully exactly what we do. Maybe we can detect if we're in wasm and disable some targets if we don't have symbols for getting entropy linked? Not sure :)

**Comment by real-or-random** [comment]
Hm, I don't think we should disable anything. This all depends on the threat model of the application and it's hard for a library to know this. (It's already hard for an application to make a guess about the threat model of the user...) 

So randomization is defense in depth against timing attacks and it may provide some resistance against other side-channel attacks (electromagnetic -- but that's anyway hard). How much do you care about this in WASM? We can't tell as a library. Anyway, can WASM code even be made resistant against timing leaks? [Apparently yes.](https://cseweb.ucsd.edu/~dstefan/pubs/watt:2019:ct-wasm.pdf) So okay, without entropy, we lose defense in depth but there's no reason to believe that we lose security. 

> Especially in wasm32-unknown-unknown you have no syscalls or way of getting entropy, so this could be problematic.

You could try to hash your secrets at least. How do you get your secrets into WASM? (I mean you can't generate them in WASM without external entropy?)

**Comment by elichai** [comment]
If you know your default implementation of `getrandom` either sucks or panics, then you should make your own global context (or use a local one) and randomize it using whatever randomness you do have (and you have to have some kind of randomness as you have a private key).

I'd rather we won't change anything in the context until https://github.com/bitcoin-core/secp256k1/pull/988 is done and then we can rehaul our API

**Comment by apoelstra** [comment]
It is really unfortunate that `rand` methods would panic. This seems like an abuse of `panic` and I don't see any reasonable way we can detect it and recover.

Are there replacements for `rand` out there which are widely supported?

**Comment by TheBlueMatt** [comment]
This is the exact reason for `global-context-less-secure` - supporting environments where you can't get global randomness, especially WASM. In these environments there's probably nothing we can do better around global context, we have to hope people use their own local contexts and randomize them manually, if they want. A "better `rand`" here would still be one that just doesn't get any randomness.

**Comment by apoelstra** [comment]
Gotcha. So, in light of #388 (not sure if you've been following this) where we are moving toward removing `global-context-less-secure` and just using `rand` whenever it was available, I guess we **can't** do that, because `rand` sometimes unpreventably terminates the program from inside library functions.

So I guess we should keep `global-context-less-secure`, indicate in the docs that this is really only if you expect `rand` to misbehave, and we should gate the global context randomization on `#[cfg(all(feature = "rand", not(feature = "global-context-less-secure"))]`.

**Comment by JeremyRubin** [comment]
you can also test rand via https://doc.rust-lang.org/std/panic/fn.catch_unwind.html at runtime btw.

**Comment by TheBlueMatt** [comment]
That is std-only, sadly (not to mention very poor practice, IMO)

**Comment by apoelstra** [comment]
@TheBlueMatt we can't prevent parallel dependencies from enabling the `rand` feature ... I don't think we can reasonably say "this may break *unless* feature X is off"

**Comment by tcharding** [comment]
I don't think we need #385. Since we feature gate on "rand-std", and not on "rand", we _do_ fully control the feature gating. If we were to use "rand" we would be at the mercy of other dependencies enabling "rand" as you say @apoelstra.

no-std users will not be enabling "rand-std" so they will not get the global context randomization and will not hit the rand panic issue, right?

**Comment by apoelstra** [comment]
Yeah, I think you're right. Ok.

**Comment by Kixunil** [comment]
Skipping over the details, would it be reasonable to provide a method for consumers to dynamically set a global randomization `Fn`? People could set it when initializing and not lose any security.

**Comment by apoelstra** [comment]
That sounds reasonable to me. @elichai what do you think?

**Comment by elichai** [comment]
> That sounds reasonable to me. @elichai what do you think?

It can work, but I doubt most rust users will understand how to use this correctly.
I do think that the best solution is to keep using `getrandom` and that if a users uses wasm32-unknown-unkown they would have to depend on `getrandom` by themselves and enable wasm-bindgen / js-sys depending on what they use.

And if they use a weirder environment then it should probably be supported in `getrandom`.

But, if it is common to use various exotic environments and they can't all be supported in `getrandom` then we could do something like that. but this sounds a bit weird to me.

FWIW there was a discussion on providing a `getrandom`-like interface in libstd: 
https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/OS.20secure.20randomness
https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/possible.20getrandom.20interface

**Comment by Kixunil** [comment]
> I doubt most rust users will understand how to use this correctly

I doubt most users will *not*. :) I imagine this doc like this

> The function is provided for platforms that don't have inherent access to randomness but require some additional bridging.
> Wasm is a popular example of such platform. It doesn't provide API to get randomness but in browsers you can still get randomness using JS bindings.
>
> **Warning:** do **not** implement custom randomness function, only bridge with existing randomness APIs on your platform. Make sure that if the API fails the supplied function fails as well (no fallbacks)!

People knowledgeable enough to use exotic platform should be knowledgeable enough to write bindings.

@TheBlueMatt 

> use global-context without the rand crate and randomize yourself 

This may be too annoying to work with. While I in general don't like global mutable state, this looks like a very reasonable exception.

**Comment by Kixunil** [comment]
Alternative idea: depend on some CSRNG (e.g. from rand crate) and only accept seed. Less bothering with creating trait objects or so. May tempt people to supply all zeroes.

**Comment by Kixunil** [comment]
My understanding of your previous comment is that you suggest consumers write something like this:

```rust
let sig1 = key.sign(ctx, msg1);
ctx.rerandomize(rng);
let sig2 = key.sign(ctx, msg2);
ctx.rerandomize(rng);
let sig3 = key.sign(ctx, msg3);
ctx.rerandomize(rng);
```

While my suggestion is to globally set a custom RNG (presumably using JS bindings) so that `sign` method automatically retrieves entropy from it and re-randomizes, so the code would become:

```rust
set_rng(my_rng);
let sig1 = key.sign(ctx, msg1);
let sig2 = key.sign(ctx, msg2);
let sig3 = key.sign(ctx, msg3);
```

Which is shorter, cleaner and thus I believe better achieves your goal of encouraging re-randomization. (I agree with the goal BTW.)

**Comment by Kixunil** [comment]
We could make it thread-local, internally using `RefCell`, or just make it non-`Send` and use interior mutability.

**Comment by Kixunil** [comment]
@TheBlueMatt hmm, looks like those platforms are screwed already unless they turn *off* re-randomization. I take back my suggestion to make context non-`Sync` (yep, that's what I meant) - consumers can already do it themselves by using `RefCell` and forcing everyone to use it just hurts modularity. If `RefCell` becomes popular, we could just write additional wrapper. (And maybe we can use `UnsafeCell` instead?)

**Comment by apoelstra** [comment]
I recently learned that `getrandom` has a `js` feature that can be enabled if we know that we're compiling for WASM in a browser context where it has access to strong randomness. I believe that if we directly depend on `getrandom` rather than going through `rand`, and expose the `js` feature through this library, that might make everyone happy.

There remains the question of what to do when the target arch is wasm, `getrandom` is enabled but `js` is not enabled. I am tempted to issue a compilation error in this case.

**Comment by JeremyRubin** [comment]
what if we have rand stay an optional dep, and have js also an optional dep on getrandom?


+1 on compiler error as mentioned!

**Comment by apoelstra** [comment]
I think we need `rand` to stay on as an optional dep, yeah, because we expose its traits in other parts of the API.

So what I'm proposing is:
* 3 features `rand` `getrandom` `js`
* `rand` implies `getrandom`
* `js` implies `getrandom/js` (and therefore implies `getrandom`, there is a cargo bug about this)

**Comment by JeremyRubin** [comment]
concept ack.

one note though -- we should be sure to not assume that all WASM contexts can have entropy through the getrandom binding, as oftentimes wasm won't have entropy at all.

**Comment by apoelstra** [comment]
This assumption is exactly what the `js` feature does.

But anyway it doesn't matter for the purpose of context rerandomization.

**Comment by Kixunil** [comment]
There's still a problem that if we want contexts to cease to be arguments and have them randomized the only way to do that is using thread locals. And `wasm` doesn't have thread locals.

But it just occurred to me that we could have a special API:

```rust
/// Opaque storage
pub struct ThreadLocal {
    ctx: Secp256k1<Sign>,
};

impl ThreadLocal {
    /// Initializes the thread-local context.
    ///
    /// Should only be used by the callers of `set_thread_local_provider.
    pub fn new() -> Self {
        // ...
    }
}

pub fn set_thread_local_provider(provider: fn() -> &'static mut ThreadLocal);
```

Applications that are guaranteed to be single-threaded can just store it in `static UnsafeCell<ThreadLocal>` and return mutable reference.

It needs some `Send`/`Sync` polishing but I don't have time to think about it RN.

**Comment by Kixunil** [comment]
> There is a proposal to add threads, but presumably if that happens thread-locals will then be supported.

Exactly and if we just use `static` assuming there is only one thread then the moment they are added the library becomes unsound because we can't predict which exact version it will be in to conditionally switch it off. Better leave this responsibility to users.

**Comment by JeremyRubin** [comment]
Question: are Contexts fork-safe?

e.g., would the following kind of thing pose issues?

```rust
let secp = Secp256k1::new();
let sk = ...;
let pid = fork();
let sk2 = new_secret_key_fn(&secp);
if  pid == 0 {
    sign(&secp, sk, "Example A");
} else {
    sign(&secp, sk, "Example B");
}
```

Despite not being threaded, I think the primitives for forking in WASM are likely to show up (with a shared memory thing). One thing that might be good to do is to ensure that it's safe to directly copy a context, or figure out how to make re-randomizing live in the host entirely (e.g., we could make a secp context that lives in the host's memory and is only available per-instance, such as follows:

```rust
struct SpecialContext<T>;

impl<T> Deref for SpecialContext<T> {
  type Target = Secp256k1<T>;
  fn deref(&self) -> &Secp256k1<T> {
    binding_to_function_that_asks_host_for_rerandomization_or_our_copy()
  }
}

```


)

**Comment by Kixunil** [comment]
> For non-wasm targets we'll need to use some sort of static mutex anyway.

Not really, thread local would be more performant without any obvious issue. But you have a point: we could default to mutex unless one activates the feature and the feature is documented to not work on wasm where atomic is used. I'd even consider not blocking in wasm: if the mutex is locked just panic.

I feel like there could be even better trick but not sure.

Also note we can't make a true mutex in `no_std` - there are no parking primitives.

@JeremyRubin if you're using `fork()` this way you probably have different problems. But yes, they are safe, just lose protection. We could register `at_fork` callback to re-randomize them though.

I'd be surprised if fork got into wasm. They specifically don't want dubious features like that one.

**Comment by JeremyRubin** [comment]
Fork like behavior already exists and is a standard thing a user might do.

https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Memory

So it's not something in spec, just something a user can do.

**Comment by apoelstra** [comment]
Fair enough -- we could make it a thread local then, but then we're back to the "maaybe wasm one day starts supporting threads without warning us" situation. Which honestly, I don't think is that big a deal -- it'll be a long time before browsers support it and a long time before people try using it, and we'll have warning.

**Comment by Kixunil** [comment]
[Please, no spin locks](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html). If we really have to lock something we should at least provide some interface to park the thread. Yes, I'm aware we have a spinning lazy initialization for `no_std` and that is also something I'd like to address eventually. I just didn't have the time for it.

**Comment by Kixunil** [comment]
Oh, yes, what you describe works for wasm (although I'd consider panicking instead of spinning) but the same problem occurs on embedded without OS (HW wallets) and we can't use that trick there so the interface to set some thread local provider is still required and once it's there we may as well enable it for wasm.

**Comment by Kixunil** [comment]
Firstly, one could wish to write a Rust library with C API linked to other languages and AFAIK it is (usually?) `no_std` to not require whole Rust std to be linked. Maybe you've heard of such project. ;)

Secondly, even if there are no `no_std` multicore platforms I don't think we can reliably detect them and refuse to compile. AFAICT having an API to provide appropriate handlers is the most practical approach and as a side effect it also supports all niche platforms without us having to do extra work.

**Comment by Kixunil** [comment]
> Well the initial discussion here was around WASM

Indeed and I point out that whatever solution works for embedded will likely work well for wasm as well. :)

Embedded platforms also have interrupts and if someone *accidentally* calls a `secp256k1` function from an interrupt spinlock means a guaranteed deadlock if it's hit.

> the user asked us to work without threading, why are we panicking

Because the user said the platform is single threaded but it's not.

> detected and compile-error'd if std isn't enabled

I believe this would cause FPs in case of libraries but it's OK if the compile error is "missing symbols for synchronization functions".

> error if neither really-no-std nor std is selected

As I mentioned in the issue at `rust-bitcoin` these almost-exclusive features are PITA, I'd rather not have them if possible.

If requiring embedded/WASM users to provide these synchronization primitives is too much boilerplate then maybe have `cfg`s (not feature - libraries shouldn't mess with it) `secp256k1-fallback-spinlocks` and `secp256k1-assert-single-thread` to set basic implementations provided by us but it'd still be useful to allow custom synchronization implementations.

**Comment by Kixunil** [comment]
FYI, `thread_local!` seems to work fine in WASM but it does require `std` which is not great. I remember there being some issue around it but it seems to be just something wrong with the `rand` crate. So maybe we could have it for those who don't mind `std` being linked to their modules.

**Comment by Kixunil** [comment]
> This isn't a spinlock issue but rather a lock issue.

The issue has a solution: disabling interrupts and checking it's not called from one. By hard-coding spinlock we are denying the users this solution.

> No, the user said the platform didn't have pthread.

Ah, yes. With pthread we can do it nicely e.g. by using its own TLS.

> We could have a different feature for "No threads, I promise", I suppose, but that's really not practical if secp is a transitive dependency.

This would work if it was `cfg` and people are supposed to only set it if they can guarantee the platform is single-threaded - e.g. current version of WASM.

> so this "support" isn't what you think it is.

Actually, it is and it is exactly what I want: have `static` `UnsafeCell` on WASM while no threads exist there and switch to actual thread locals once they do (and the compiler/`std` is responsible for this). This way the library is sound.

**Comment by Kixunil** [comment]
> the only solution is to make the user pass the context

Not really, instead of passing a context to each individual function you could supply a custom "thread local" provider. E.g. disable interrupts  and check it's not being run in one on embedded, return a reference to static on WASM...

```rust
// libsecp256k1
#[cfg(secp256k1_custom_thread_local)]
pub mod custom_thread_local {
    pub struct ThreadLocal(Context); 

    impl ThreadLocal {
        /// Initializes the thread-local data.
        ///
        /// This function is unsafe to implement and MUST NOT call secp256k1_get_thread_local nor secp256k1_get_thread_local
        pub fn new() -> Self { /* ... */ }
    }

    extern "Rust" {
        /// Provides a pointer to secp256k1-owned thread local.
        ///
        /// The implementor must return a valid pointer to `ThreadLocal` obtained by calling `ThreadLocal::new()`.
        /// The pointer must be valid until the caller calls `secp256k1_release_thread_local` or the current thread exists whichever occurs first.
        /// The caller MUST NOT dereference the pointer afterwards.
        /// The function is NOT reentrant: the caller may only call it again if it called `secp256k1_release_thread_local` first.
        ///
        /// Note that if this is called from an interrupt the implementor either has to return a separate instance or panic.
        /// If the interrupt itself can be interrupted it's probably best to just panic since each call would have to return a new, different instance
        /// and doing cryptographic operations in such interrupts is most likely a bad idea anyway.
        ///
        /// The function is both unsafe to call and unsafe to *implement*.
        #[no_mangle]
        pub unsafe fn secp256k1_get_thread_local() -> *const ThreadLocal;

        /// Called when the caller of `secp256k1_get_thread_local` no longer needs access to the thread local.
        ///
        /// This is usually a no-op but some platforms may implement "thread locals" as globals behind mutex if there is no better way or, on embedded, these may be implemented as disabling and re-enabling interrupts. (But see note above.)
        #[no_mangle]
    pub unsafe fn secp256k1_release_thread_local();
}

/// Provides safe access to the thread local.
pub(crate) fn with_thread_local<R, F: FnOnce(&ThreadLocal) -> R>() ->R {
    struct Guard;
    impl Drop for Guard {
        fn drop(&mut self) {
            unsafe { secp256k1_release_thread_local() }
        }
    }

    let ptr = secp256k1_get_thread_local();
    let _guard = Guard;
    fun(unsafe { &*ptr })
}
```

The root crate then sets the `--cfg secp256k1_custom_thread_local` and provides the symbols on less common platforms.

> If you take existing code and compile for WASM, it'll work great in a threading context, just be randomly unsafe and panic-y.

I've read that there is some kind of access control to the memory so threads can't access the memory of each other unless specifically marked as shared which happens during compilation. Maybe my understanding is wrong though and if you can recommend good resources on the topic I'd love to learn!

**Comment by Kixunil** [comment]
> is just going to be missed

Compiler errors are pretty hard to miss, I think.

>  response is going to be to switch to another library, not fix it.

To which library? AFAIK there isn't one with this level of rigor. The thing here is unfixable. The best alternative we could do is add them as normal features and document that library crates must not depend on them but I suspect some will not notice it.

> but the real solution is to just not do cryptographic operations in an interrupt context

It is but the crate has to remain sound and panic is better than deadlock. But it just occurred to me that even simpler embedded implementation is `if is_this_interrupt() { panic!("attempt to do crypto in interrupt"); } else { STATIC_CTX.ptr() }`. The problem is we can't implement `is_this_interrupt` so we still need user to provide the checking function or just simply ban reentrancy but that should not be the default.

> rustc will need a whole new target for wasm-with-locking?

IIRC currently there is `-C target-feature=+atomics` (std still needs to be rebuilt)

>  you can't combine wasm-threaded and non-wasm-threaded code at all?

Perhaps you can if they communicate over something like sockets but they can't access each-others memory.

**Comment by Kixunil** [comment]
> unless you require all downstream crates to provide an implementation

Just the root crate. If there isn't any of the listed `cfg`s **and** the crate is `no_std` we emit `compile_error!`, with `cfg` we can do some limited things (linking to `pthread`) or just let users implement it themselves.  And if there are popular implementations they can be turned into a crates so people don't have to reinvent the wheel.

> We've seen many many times people use a number of other crates that are much less rigorous to work around various usability issues.

As I said, I want an example to be able to judge that this kind of usability issue would be a likely to push people away.

> If it deadlocks users will see it just as much as they'd see a panic

There is a crazy huge difference between an app being randomly stuck (which of the dependncies caused it? How do I even get a stacktrace on embedded?) and a clear message "You called x from an interrupt which is forbidden". I've spent days with a colleage debugging deadlocks in a *desktop* application which is already easier than embedded. Reading a message takes two seconds, not days, which makes it literally tens of thousands times faster to debug.

> just happens to work correctly sometimes in addition.

"Works correctly sometimes" (Usually in test env) and then randomly deadlocks (usually in production) is the worst thing you can get. A reliably crashing application will not get past QA in any serious company, so the end users won't see anything. With a clear message the developers will spend much less time figuring out what's wrong.

> eg I can link with C code with clang and break it, it seems

I guess, interesting. From what I've read they plan to improve it but it's difficult because of poor WASM threading spec.

**Comment by Kixunil** [comment]
> C directly without a static context

So you provide the context in each call? Isn't it more annoying than having to activate a special feature once?

> to force downstream users to implement some TLS logic just because they want to use no-std.

I wouldn't mind keeping the current functions requiring the context argument. :man_shrugging: Sadly, I think library crates will be lazy and not use them.

>  to implement some TLS logic

I don't find "implement" an accurate description. More like "If you are on platforms x, y, z you can activate this feature and not worry about it, if you are on embedded without OS you need to provide the implementation or check these community crates providing it for common MCUs."

> debug feature for "panic if you see threading", fine?

If it's considered a debug feature then perhaps also record `core::panic::Location` of the caller currently using the context and use `track_caller`. :) And have a warning about potential deadlocks at the appropriate place. (still "which of my 10 dependncies deadlocked?" is a problem, maybe unsolvable)

> the "default" absolutely must work totally fine without effort

I don't think this is possible in general if they also want static context. But if there is still a way to get randomness we could do:

```rust
if GLOBAL_CTX_IS_USED.compare_exchange(false, true, Acquire, Relaxed) {
    let mut ctx = Secp256k1::new_randomized(GLOBAL_RNG.get_32_b());
    do_something_with_ctx(&mut ctx);
} else {
    let set_unused_on_drop = SetUnusedOnDrop; // guard doing `GLOBAL_CTX_IS_USED.store(Release)` on drop
    // SAFETY: we just checked nothing else uses this
    do_something_with_ctx(unsafe { &mut *GLOBAL_CTX.get_mut() })
}
```

Then the default is a performance hit as you proposed re-creating randomized context if there are actually "threads". Maybe we could log a warning somehow though.

> SGX

Oh, interesting, I will have to check their API/docs to see what's possible. Do you have some recommended reading material?

> I believe rustc is back to setting noalias on mut references, which is incorrect if you have an interrupt, no?

It's only incorrect if the said interrupt is creating a mutable reference to the same memory. This is already banned in Rust anyway and it's why you have to go through `UnsafeCell` and implement some synchronization if you actually want two things to hold those references - e.g. disabling interrupts in case of embedded.

> writes a bunch of delicate unsafe code

... or sets a correct flag depending on their platform.

> calling cryptographic operations in parallel from a single thread

That's not even physically possible. :) Our static-context functions will not call other static-context functions (while holding `&mut ThreadLocal`)

**Comment by apoelstra** [comment]
This is a fascinating conversation but I'll just remind people that randomization is a defense-in-depth feature that is not needed for security or correctness, and my stance is that we should **just disable it for wasm** and maybe expose a `js` feature to recover it if users think the js rng is usable and ok.

I think this has gotten a bit past WASM to a general "platforms without TLS" ... okay, then if we have std let's use TLS and if we don't we'll just disable randomization on the static context.

**Comment by apoelstra** [comment]
> Yea, I do kinda wonder if there isn't a decent way to get this without relying on `rand`. Its kinda unclear to me what the "right" API would be for "no, I do really have a RNG, but I also don't have TLS", but I do think we should at least try to build that?

If it's purely the `rand` dependency then we can use the secret key as entropy and only support re-randomizing when signing (which, arguably, is the only time that we need to rerandomize). Maybe we should make this change, to at least deal with the current wasm/js issue.

But this still leaves open the issue of having a mutable context at all, and what synchronization we need to layer onto that.

**Comment by apoelstra** [comment]
>ACK. Just to double-check, there's no concern with the randomization process itself being non-constant-time and leaking bits? Should we instead consider randomizing with the signature itself (or is it useless given the signature is public?)

We should take a tagged hash of the secret key -- the concern is not constant-timeness but rather that the re-randomization internally computes a curvepoint (i.e. the public key) and so the resulting re-randomization wouldn't be a secret. And yes -- to be effective the rerandomization must be secret, so we can't use the signature itself.

In our discussions about doing this upstream we observed that when generating a signature, there is a secret bit (the original parity of `R`) that we just throw away, which we could use to do one bit of re-randomization per signature. Then unless we had such an egregious sidechannel that an attacker could learn information from observing only a couple signatures, we'd get a very fast form of automatic re-randomization.

The hold-up there is (a) upstream having the same kind of API fights that we're having, re the signing function mutating the context object, and (b) trying to formalize exactly what we hope to accomplish, to understand whether 1-bit of rerandomization is effective.

>Heh, now we're back where we started. I still favor a trivial spinlock for no-std with a feature or two to debug deadlocks or provide custom TLS.

I tend to agree with you, but I see @Kixunil's point. Personally what I'd like is to just implement this with atomics ... but I need upstream support for this which would be difficult to get because of how badly it'd break their abstraction.

**One alternative we haven't considered is to use a variant of a spirlock where after spinning a couple of times we just give up and don't rerandomize.** @Kixunil how would you feel about that? This would prevent any possibility of deadlocking, still work in the vast majority of cases. We could even add another atomic to allow run-time configuring the number of times to run before giving up, though it's hard for me to imagine anybody understanding or using that capability..

**Comment by Kixunil** [comment]
> after spinning a couple of times we just give up and don't rerandomize.

This is quite interesting! It's still silently doing things one may not want so I'd still prefer to give debug options and ideally also thread local API. I was burnt badly by code silently swallowing errors way too many times to just ignore it. I wonder if it'd be acceptable to have an optional dependency on the `log` crate and logging warnings. It has a very conservative MSRV policy and is maintained by the `rust-lang` team.

Regarding the number of spins I would actually set it to zero. If there is contention on presumably single-threaded platform then the culprit is most likely reentrancy and there's no point in spinning. Note that I still want `pthread` or `std` or whatever means to get TLS and this should be used as last resort only.

Also anything surprising like this should be very visibly documented so that anyone who requires randomization knows about it. And ideally I'd love to see some "don't compile dangerous configurations" option somehow. This doesn't help if the crate is a dependency of a dependency but perhaps the consumers who don't audit don't care about this anyway, so it's OK?

**Comment by apoelstra** [comment]
If `log` is useful by itself then I wouldn't mind having an optional dependency on it. IIRC you usually need some other crate to actually use it and those other crates are a bit less stable.

>Regarding the number of spins I would actually set it to zero. If there is contention on presumably single-threaded platform then the culprit is most likely reentrancy and there's no point in spinning. Note that I still want pthread or std or whatever means to get TLS and this should be used as last resort only.

SGTM.

>Also anything surprising like this should be very visibly documented so that anyone who requires randomization knows about it.

Sure, we can document it, though even as an author of libsecp I'm having trouble imagining anybody who would require randomization..

>And ideally I'd love to see some "don't compile dangerous configurations" option somehow

We can do this with an additional `cfg` param, but with cargo features I think it'd be too disruptive since it would prevent people composing dependency trees.

**Comment by Kixunil** [comment]
>  I'm having trouble imagining anybody who would require randomization..

Oh, doesn't it imply it's not actually valuable enough to have it?

>  additional `cfg` param, but with cargo features I think it'd be too disruptive

Indeed. It's a shame this wasn't addressed in Cargo yet.

**Comment by apoelstra** [comment]
@TheBlueMatt we're no longer talking about panicking, but instead just not randomizing.

>Oh, doesn't it imply it's not actually valuable enough to have it?

I mean, we should support it when it's low-effort. It's defense in depth. I can't quantify its value. My point is that nobody else can quantify it either, so nobody "needs" it, and we definitely are not going to have people using such obscure options that even I can't tell what their purpose is.

**Comment by Kixunil** [comment]
> deadlocking is not "silently swallowing errors"

I meant the library compiling without complaining but:

> we should support it when it's low-effort.

is a nice way to think about it and I can finally understand it better. Although the question is "what considered low-effort?"
`std` with thread local definitely yes, `pthread` - IDK, maybe? We need to detect it being available and I suspect the only reliable way is to attempt to compile a trivial piece of code using it. No-rerandomize fallback definitely. A feature to debug it also. Thread local extern interface looks simple enough to me as well - trivial on our side and trivial for consumers whose platforms provide `thread_local(Key) -> *const X`.

For logging, I guess it could be used elsewhere too. That'd make it more justified.

**Comment by apoelstra** [comment]
IMHO "low-effort" means
* with `std` we use TLS
* without `std` we use the "sorta-spinlock which skips randomization if there is any contention at all"

**Comment by Kixunil** [comment]
OK, I on board with doing this (with appropriate documentation; including logging) initially and leaving the rest until someone actually asks for more.

**Comment by apoelstra** [comment]
So I don't forget: this library is std-enabled and has several deps we don't want, but has several tricks in its source code for extracting entropy from CPU jitter and nanosecond-scale timing and stuff, which be sufficient for context rerandomization purposes even on wasm.

https://github.com/DavidVorick/userspace-rng/
